blog/_posts/2017-01-24-cudanative-nvidia.md

```diff
@@ -0,0 +1,234 @@
+
+But how does it perform? The chart below shows the minimum execution time of both
+implementations as measured by [this benchmark
+script]({{page.cudanative_tree}}/examples/reduce/benchmark.jl). The code generated by
+CUDAnative.jl performs well for a large problem size, but there seems to be a constant
+overhead (note the logarithmic scale). Luckily, there is an easy explanation: we haven't
+fully implemented the NVVM ABI yet, and pass aggregate objects like the arrays in this
+example by pointer. This requires extra memory allocations and copies, which synchronize the
+GPU and slow down the kernel launch sequence.
+
+![Performance comparison of parallel reduction
+implementations.](/images/blog/2017-01-24-cudanative/performance.png)
+
+We also aim to be compatible with existing tools from the CUDA toolkit. For example, we
+generate the necessary line-number information for the NVIDIA Visual Profiler to [work as
+expected](/images/blog/2017-01-24-cudanative/nvpp.png), and [wrap relevant API
+functions]({{page.cudanative_tree}}/src/profile.jl) to have more fine-grained control. The
+line-number information also enables accurate backtraces in combination with tools like
+`cuda-memcheck`:
+
+```
+$ cuda-memcheck julia examples/oob.jl
+========= CUDA-MEMCHECK
+========= Invalid __global__ write of size 4
+=========     at 0x00000148 in examples/oob.jl:14:julia_memset_66041
+=========     by thread (10,0,0) in block (0,0,0)
+=========     Address 0x1020b000028 is out of bounds
+```
+
+Full debug information is not supported by the LLVM NVPTX back-end yet, so `cuda-gdb` and
+friends will not work as well.
+
+<!-- TODO: make sure profile.jl contains some more impressive code -->
+
+
+### Future developments
+
+The combination of carefully-designed language semantics and a specializing JIT compiler
+makes it possible to create abstractions without paying the price. CUDAnative.jl aims to
+extend this to the GPU and make it possible to create high-performance GPU abstractions in a
+high-level language. We aren't quite there yet — certain language features are not supported
+or well-optimized, the standard library still needs to be integrated — but interesting new
+developments are already appearing.
+
+For example, the [GPUArrays.jl](https://github.com/JuliaGPU/GPUArrays.jl) package is a
+playground for a vendor- and toolkit-agnostic array abstraction. It implements Julia's
+higher-order functions, like `map` and `broadcast`, compiling user-code for the GPU in use.
+
+Another powerful demonstration of Julia's capabilities is the proof-of-concept
+[CUDAnativelib.jl](https://github.com/JuliaGPU/CUDAnativelib.jl), which makes it possible to
+use device libraries from Julia GPU functions. It uses
+[Cxx.jl](https://github.com/Keno/Cxx.jl) to parse or even compile CUDA C header files,
+building on Julia's powerful foreign function interface to generate the required calling
+sequences.
```

Comment:
`This will allow us to take advantage of the native device libraries, like cuRAND, that NVIDIA provides.` or something to that extend.